### PR TITLE
More restrictive tslint and more Ember definitions.

### PR DESCRIPTION
**Pull request by winding-lines**
Please fill in this template.

- [x] Prefer to make your PR against the `types-2.0` branch.
- [x] Test the change in your own code.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).
- [x] Run `npm run lint -- package-name` if a `tslint.json` is present.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: http://emberjs.com/api/classes/Ember.inject.html

- fix a lot of lint issues
- add inject
- enhance Promise<T, V>

### CODE REVIEW

**Comment by winding-lines** [comment]
@jedmao let me know what you think :)

File: ember/index.d.ts [hunk 1]
```diff
@@ -2557,230 +2490,22 @@ declare namespace Ember {
...
-    var wrap: typeof Ember.wrap;
-    var _ContainerProxyMixin : typeof Ember._ContainerProxyMixin;
-    var _RegistryProxyMixin: typeof Ember._RegistryProxyMixin;
-    var getOwner: typeof Ember.getOwner;
-    var setOwner: typeof Ember.setOwner;
-    var testing: typeof Ember.testing;
-    var MODEL_FACTORY_INJECTIONS: typeof Ember.MODEL_FACTORY_INJECTIONS;
-    var assign: typeof Ember.assign;
-}
+import Em = Ember;
```
Comment (jedmao) [inline ember/index.d.ts]: Neat trick!

**Review by jednano — APPROVED** [review]
Nice work!

**Comment by jedmao** [comment]
👍

**Comment by winding-lines** [comment]
Thanks Jed! You did an amazing work  on the Ember type definition.

Marius
---

**Review by jednano — APPROVED** [review]

**Pull request merged**